File: packages/pip/src/goodvibes_cli/steps/write_manifest.py
from __future__ import annotations

import hashlib
import json
import pathlib

from goodvibes_cli.utils.json_merge import write_json
from goodvibes_cli.utils.safe_path import check_writable

MANIFEST_PATH = ".goodvibes.json"
# Not hex digests: "user-owned" never matches a file's hash, "user-removed" marks a file the user deleted.
USER_OWNED = "user-owned"
USER_REMOVED = "user-removed"


def posix_key(rel: str) -> str:
    # Windows runs used to write backslash keys; one spelling keeps every comparison exact.
    return rel.replace("\\", "/")


def write_manifest(
    dest_dir: pathlib.Path,
    written_files: list[str],
    version: str,
    preserved: dict[str, str] | None = None,
    managed: dict[str, list[str]] | None = None,
    scope: str | None = None,
    git_hook: str | None = None,
) -> None:
    # Preserved hashes come only from the prior manifest, never re-read from dest,
    # so a skipped (user-modified) file can't be silently reclassified as unmodified.
    files: dict[str, str] = {posix_key(k): v for k, v in (preserved or {}).items()}
    for rel in written_files:
        content = (dest_dir / rel).read_bytes()
        files[posix_key(rel)] = hashlib.sha256(content).hexdigest()
    manifest: dict = {"version": version, "files": files}
    if managed is not None:
        manifest["managed"] = managed
    if scope is not None:
        manifest["scope"] = scope
    if git_hook is not None:
        manifest["gitHook"] = git_hook
    check_writable(dest_dir, dest_dir / MANIFEST_PATH)
    write_json(dest_dir / MANIFEST_PATH, manifest)


class ManifestError(ValueError):
    pass


def read_manifest(dest_dir: pathlib.Path) -> dict | None:
    p = dest_dir / MANIFEST_PATH
    if not p.exists():
        return None
    fix = "fix it or delete it and run goodvibes init"
    try:
        data = json.loads(p.read_text(encoding="utf-8"))
    except ValueError as e:
        raise ManifestError(f"{p} is not valid JSON ({e}); {fix}") from e
    if not isinstance(data, dict):
        raise ManifestError(f"{p} is not valid JSON (expected an object, found {type(data).__name__}); {fix}")
    if isinstance(data.get("files"), dict):
        data["files"] = {posix_key(k): v for k, v in data["files"].items()}
    return data

File: packages/npm/src/steps/write-manifest.ts
import { createHash } from 'node:crypto'
import { readFile } from 'node:fs/promises'
import { printable, writeBlocked, writeFileAtomic } from '../utils/fs-safe.js'
import { isJsonObject } from '../utils/json-merge.js'
import { join } from 'node:path'

export interface Manifest {
  version: string
  files: Record<string, string>
  managed?: Record<string, string[]>
  scope?: 'global' | 'project'
  gitHook?: 'installed' | 'user-removed'
}

export const MANIFEST_PATH = '.goodvibes.json'

// Sentinels, not hex digests: 'user-owned' is never overwritten; 'user-removed' was deleted by the user and is not re-added.
export const USER_OWNED = 'user-owned'
export const USER_REMOVED = 'user-removed'

// Manifest keys are always forward-slash so a manifest written on Windows matches on every OS.
export const posixKey = (rel: string): string => rel.replace(/\\/g, '/')
const posixKeys = <T>(record: Record<string, T> = {}): Record<string, T> =>
  Object.fromEntries(Object.entries(record).map(([k, v]) => [posixKey(k), v]))

export async function writeManifest(
  destDir: string,
  writtenFiles: string[],
  version: string,
  preserved?: Record<string, string>,
  managed?: Record<string, string[]>,
  scope?: 'global' | 'project',
  gitHook?: Manifest['gitHook'],
): Promise<string | null> {
  const blocked = await writeBlocked(destDir, MANIFEST_PATH)
  if (blocked) return blocked
  // Preserved hashes come only from the prior manifest, never re-read from dest,
  // so a skipped (user-modified) file can't be silently reclassified as unmodified.
  const files: Record<string, string> = posixKeys(preserved)
  for (const rel of writtenFiles) {
    const content = await readFile(join(destDir, rel), 'utf-8')
    files[posixKey(rel)] = createHash('sha256').update(content, 'utf8').digest('hex')
  }
  const manifest: Manifest = { version, files, ...(managed ? { managed: posixKeys(managed) } : {}), ...(scope ? { scope } : {}), ...(gitHook ? { gitHook } : {}) }
  await writeFileAtomic(join(destDir, MANIFEST_PATH), JSON.stringify(manifest, null, 2) + '\n')
  return null
}

// Keys reach delete and write calls, and a cloned repo can ship its own manifest: "a/../../.git/HEAD" must never pass a prefix check.
const unsafeKey = (k: string): boolean =>
  /[\u0000-\u001f\u007f-\u009f]/.test(k) || /^[A-Za-z]:/.test(k) || k.split('/').some(s => s === '' || s === '.' || s === '..')

const mapOf = (v: unknown, ok: (x: unknown) => boolean): boolean => isJsonObject(v) && Object.values(v).every(ok)

function manifestProblem(m: Record<string, unknown>): string | null {
  if (m.files !== undefined && !mapOf(m.files, v => typeof v === 'string')) return '"files" is not a JSON object of file paths to hashes'
  if (m.managed !== undefined && !mapOf(m.managed, v => Array.isArray(v) && v.every(s => typeof s === 'string'))) {
    return '"managed" is not a JSON object of file paths to lists of text'
  }
  if (m.scope !== undefined && m.scope !== 'global' && m.scope !== 'project') return '"scope" is not "global" or "project"'
  if (m.gitHook !== undefined && m.gitHook !== 'installed' && m.gitHook !== USER_REMOVED) return '"gitHook" is not "installed" or "user-removed"'
  const bad = [...Object.keys(m.files ?? {}), ...Object.keys(m.managed ?? {})].map(posixKey).find(unsafeKey)
  return bad === undefined ? null : `"${bad}" is not a safe relative path`
}

// Throws an actionable error for a manifest that exists but cannot be used; guessing would lose tracking.
export function parseManifest(raw: string, path: string): Manifest {
  const fail = (why: string) => new Error(printable(`${path} ${why}; fix it or delete it and run goodvibes init`))
  let data: unknown
  try {
    data = JSON.parse(raw)
  } catch (e) {
    throw fail(`is not valid JSON (${(e as Error).message})`)
  }
  if (!isJsonObject(data)) throw fail('is not valid JSON (not a JSON object)')
  const problem = manifestProblem(data)
  if (problem) throw fail(`is not a valid goodvibes manifest (${problem})`)
  const m = data as unknown as Manifest
  return { ...m, files: posixKeys(m.files), ...(m.managed ? { managed: posixKeys(m.managed) } : {}) }
}

export async function readManifest(destDir: string): Promise<Manifest | null> {
  const path = join(destDir, MANIFEST_PATH)
  let raw: string
  try {
    raw = await readFile(path, 'utf-8')
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === 'ENOENT') return null
    throw e
  }
  return parseManifest(raw, path)
}
